crates/bevy_ecs/src/system/commands/command_queue.rs

```diff
@@ -0,0 +1,225 @@
+/// accesses are safe.
+unsafe fn invoke_command<T: Command>(command: *mut u8, world: &mut World) {
+    let command = if std::mem::size_of::<T>() > 0 {
+        command.cast::<T>().read_unaligned()
+    } else {
+        // NOTE: This is necessary because if the `CommandQueueInner` is only filled with 0-sized
+        // commands the `bytes` vec will never allocate. Then means that `bytes.as_ptr()` could be null
+        // and reading a null pointer is always UB.
+        // However according to https://doc.rust-lang.org/std/ptr/index.html
+        // "The canonical way to obtain a pointer that is valid for zero-sized accesses is NonNull::dangling"
+        // therefore the below code is safe to do.
+        let ptr = std::ptr::NonNull::<T>::dangling().as_ptr();
+        ptr.cast::<T>().read()
+    };
+    command.write(world);
+}
+
+struct CommandMeta {
+    offset: usize,
+    func: unsafe fn(value: *mut u8, world: &mut World),
+}
+
+#[derive(Default)]
+pub(crate) struct CommandQueueInner {
+    bytes: Vec<u8>,
+    metas: Vec<CommandMeta>,
+}
+
+// SAFE: All commands [`Command`] implement [`Send`]
+unsafe impl Send for CommandQueueInner {}
+
+// SAFE: `&CommandQueueInner` never gives access to the inner commands.
+unsafe impl Sync for CommandQueueInner {}
+
+impl CommandQueueInner {
+    /// Push a new `command` onto the queue.
+    #[inline]
+    pub fn push<C>(&mut self, command: C)
+    where
+        C: Command,
+    {
+        let size = std::mem::size_of::<C>();
+        let old_len = self.bytes.len();
+
+        self.metas.push(CommandMeta {
+            offset: old_len,
+            func: invoke_command::<C>,
+        });
+
+        if size > 0 {
+            self.bytes.reserve(size);
+
+            // SAFE: The internal `bytes` vector has enough storage for the
+            // command (see the call the `reserve` above), and the vector has
+            // its length set appropriately.
+            // Also `command` is forgotten at the end of this function so that
+            // when `apply` is called later, a double `drop` does not occur.
+            unsafe {
+                std::ptr::copy_nonoverlapping(
+                    &command as *const C as *const u8,
+                    self.bytes.as_mut_ptr().add(old_len),
+                    size,
+                );
+                self.bytes.set_len(old_len + size);
+            }
+        }
+
+        std::mem::forget(command);
+    }
+
+    /// Invoke each command `func` for each inserted value with `world`
+    /// and then clears the internal bytes/metas command vectors.
+    pub fn apply(&mut self, world: &mut World) {
+        let byte_ptr = self.bytes.as_mut_ptr();
+
+        // SAFE: In the iteration below, `meta.func` will safely consume and drop each pushed command.
+        // This operation is so that we can reuse the bytes `Vec<u8>`'s internal storage and prevent
+        // unnecessary allocations.
+        unsafe { self.bytes.set_len(0) };
```

Comment:
Has this been run through miri with `-Zmiri-track-raw-pointers`?

I think under that model, this might be UB, because the parent borrow of the raw pointer `byte_ptr` finishes (because you've used `bytes` again), then `byte_ptr` is being used again.

(And this is why friends don't let friends use `unsafe`)

Reply:
The inspiration for this thought was probably https://github.com/rust-lang/rust/pull/85610

Reply:
I might be a bit confused, but I don't believe `set_len` does anything to the `Vec`'s buffer pointer?

Reply:
My understanding of this is as follows:

The signature of `set_len` is 
```rust
fn set_len(&mut self, ...)->...;
```
Which means that there's a reborrow of the entirety of `self.bytes`
However, `self.bytes`' parent in the borrow stack is the borrow of `self.bytes` for `as_mut_ptr` on line 81.
This parent borrow however is invalidated by the reborrow, and so the raw pointer is no longer valid.

If this does pass miri, then my understanding is probably wrong.

Reply:
Ahh I see. 
Basically `get_mut_ptr` is a mutable access to `bytes`. and then `set_len` is also a mutable access to bytes...
And when we call `byte_ptr.add()` we're re-accessing the ptr which is UB.

Reply:
I think 'might be UB' is more appropriate, if my understanding of the caveats around `track-raw-pointers` are correct.

But otherwise yes, exactly.

Reply:
I could possible to do some interesting iterator chaining here where we end up calling `bytes.drain(..)` and avoid this all together.

Reply:
Also wouldn't a simple solution be to just call `set_let` before `as_mut_ptr()`?

Reply:
I think that should work fine, yes.

Reply:
Interestingly after running 
```
MIRIFLAGS="-Zmiri-track-raw-pointers" cargo miri test --package bevy_ecs --lib --all-features --system::commands::command_queue::test
```
`miri` doesn't complain about anything.

Reply:
Oddly enough after running
```
MIRIFLAGS="-Zmiri-track-raw-pointers" cargo miri test --package bevy_ecs --lib --all-features --system::commands::command_queue::test
```
Miri doesn't seem to complain :/

Reply:
Also interestingly if you look here:
https://doc.rust-lang.org/std/ptr/fn.copy_nonoverlapping.html
specifically under the `Examples` you'll see the following:
```rust
        let src_ptr = src.as_ptr();

        // Truncate `src` without dropping its contents. We do this first,
        // to avoid problems in case something further down panics.
        src.set_len(0);

        // The two regions cannot overlap because mutable references do
        // not alias, and two different vectors cannot own the same
        // memory.
        ptr::copy_nonoverlapping(src_ptr, dst_ptr, src_len);
``` 
IIUC, this could possibly be UB as well? Since the call to `set_len(0)` (which is mutable) is after the call to `as_ptr()` which would theoretically break the `as_ptr()`'s immutable borrow.

Reply:
The rules are definitely subtle

They might both be fine, but the version with set_len first is probably more fine.